File: backend/api/routes/master.go
package routes

import (
	"chairTime/api"
	masterController "chairTime/api/controllers/master"
	"chairTime/constant"

	"github.com/labstack/echo/v4"
)

func MasterRoute(app *api.Application, groupRoute echo.Group) {
	adminRoleId, userRoleId, masterRoleId := constant.AdminRoleId, constant.UserRoleId, constant.MasterRoleId

	groupRoute.GET("/master/unavailables/:master_id", app.RouteHandler(masterController.GetMasterUnavailableSchedules))
	groupRoute.GET("/masters", app.RouteHandler(masterController.GetMasters), app.Authorization([]int{adminRoleId, userRoleId}))
	groupRoute.GET("/master/:master_id/bookings", app.RouteHandler(masterController.GetMasterBookings), app.Authorization([]int{masterRoleId}))
	groupRoute.GET("/master/:master_id/styles-offer", app.RouteHandler(masterController.GetMasterStylesOffer), app.Authorization([]int{adminRoleId, userRoleId}))
	groupRoute.GET("/master/:master_id", app.RouteHandler(masterController.GetMasterById))
	groupRoute.GET("/master/:master_id/:date", app.RouteHandler(masterController.GetMasterAvailableTimeSlots))
	groupRoute.PUT("/master/update", app.RouteHandler(masterController.UpdateMaster), app.Authorization([]int{adminRoleId}))
	groupRoute.PUT("/master/unavailable/update", app.RouteHandler(masterController.UpdateMasterUnavailableSchedule))
	groupRoute.POST("/master/create", app.RouteHandler(masterController.CreateMaster), app.Authorization([]int{adminRoleId}))
	groupRoute.POST("/master/unavailable", app.RouteHandler(masterController.CreateMasterUnavailableSchedule))
	groupRoute.DELETE("/master/:unavailable_schedule_id", app.RouteHandler(masterController.DeleteMasterUnavailableSchedule))
}

File: frontend/src/api/master.ts
import type { ResponseSuccess, ResponseSuccessWithData, ResponseWitMetaType } from "@/core/models/base";
import { privateHttp } from "./http";
import type { CreateMaster, CreateMasterUnavailableScheduleType, MasterStylesOfferType, MasterType, MasterUnavailableScheduleType, WorkingTime } from "@/core/models/master";
import type { MasterBookingType } from "@/core/models/booking";

export function getMasters(): Promise<ResponseWitMetaType<MasterType[]>> {
    return privateHttp.get("/masters")
}

export function getMasterStylesOffer(masterId: number): Promise<ResponseSuccessWithData<MasterStylesOfferType[]>> {
    return privateHttp.get(`/master/${masterId}/styles-offer`)
}

export function getMasterUnavailableSchedules(masterId: number): Promise<ResponseSuccessWithData<MasterUnavailableScheduleType[]>> {
    return privateHttp.get(`/master/unavailables/${masterId}`)
}

export function getMasterBookings(masterId: number): Promise<ResponseSuccessWithData<MasterBookingType[]>> {
    return privateHttp.get(`/master/${masterId}/bookings`)
}

export function getMasterAvailableTimeSlots(masterId: number, date: string): Promise<ResponseSuccessWithData<string[]>> {
    return privateHttp.get(`/master/${masterId}/${date}`)
}

export function getMasterById(masterId: number): Promise<ResponseSuccessWithData<MasterType>> {
    return privateHttp.get(`/master/${masterId}`)
}

export function updateMasterWorkingTime(payload: WorkingTime): Promise<ResponseSuccessWithData<WorkingTime>> {
    return privateHttp.put("/master/working-times", payload)
}

export function scheduleMasterUnavailableDays(payload: CreateMasterUnavailableScheduleType[], master_id: number): Promise<ResponseSuccess> {
    return privateHttp.post(`/master/${master_id}/unavailable`, payload)
}

export function deleteMaster(masterId: number): Promise<ResponseSuccess> {
    return privateHttp.delete(`/master/${masterId}`)
}

export function createMaster(masterPayload: CreateMaster): Promise<ResponseSuccessWithData<MasterType>> {
    return privateHttp.post("master/create", masterPayload)
}